src/main/resources/jupyter/init-actions.sh

```diff
@@ -109,6 +110,17 @@ if [[ "${ROLE}" == 'Master' ]]; then
       docker exec -d ${JUPYTER_SERVER_NAME} ${JUPYTER_HOME}/install-jupyter-extension.sh ${JUPYTER_HOME}/${JUPYTER_EXTENSION_ARCHIVE}
     fi
 
+
+    # If a Jupyter user script was specified, copy it into the jupyter docker container.
+    if [ ! -z ${JUPYTER_USER_SCRIPT_URI} ] ; then
+      gsutil cp ${JUPYTER_USER_SCRIPT_URI} /etc
+      JUPYTER_USER_SCRIPT_ARCHIVE=`basename ${JUPYTER_USER_SCRIPT_URI}`
+      docker cp /etc/${JUPYTER_USER_SCRIPT_ARCHIVE} ${JUPYTER_SERVER_NAME}:${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT_ARCHIVE}
+      docker exec -u root -d ${JUPYTER_SERVER_NAME} chmod +x ${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT_ARCHIVE}
+      docker exec -u root -d ${JUPYTER_SERVER_NAME} ${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT_ARCHIVE}
```

Comment:
These are going to run after the Docker entry point, so pyspark will already be running, and potentially installed things won't be picked up (they definitely won't if they're Python packages).

I think the approach you need to take is to edit the dockerfile so that instead of calling pyspark, it calls a bash script that first runs the user script and then launches pyspark. You will likely have to mess around with when the dockerfile changes user to make sure it runs as root but the Jupyter user doesn't.

Reply:
I tested this with the following script:

`#!/usr/bin/env bash

pip install arrow`

And from my notebook I could import the library and use it - so I was able to install a python package.

But - I'm fine with doing it through the dockerfile!